### PR TITLE
Pull corelib from cargo cache for path/git builds

### PR DESCRIPTION
**Pull request by maciektr**

### CODE REVIEW

File: scarb/Cargo.toml [hunk 1]
```diff
@@ -52,19 +52,19 @@ scarb-build-metadata = { path = "../utils/scarb-build-metadata" }
...
-serde_repr.workspace = true
 sha2.workspace = true
 smallvec.workspace = true
 smol_str.workspace = true
 tar.workspace = true
 thiserror.workspace = true
 tokio.workspace = true
-toml.workspace = true
 toml_edit.workspace = true
+toml.workspace = true
```
[the remark this answers is not in the document]
Comment (maciektr) [reply scarb/Cargo.toml]: This came from vscode, indeed intellij does it another way.

File: utils/scarb-build-metadata/build.rs [hunk 2]
```diff
@@ -86,6 +91,31 @@ fn cairo_version() {
...
+        // Crate root directory.
+        .parent()
+        // The `crates` directory from Cairo repository.
+        .and_then(|p| p.parent())
+        // The Cairo repository root.
+        .and_then(|p| p.parent())
+        // Corelib should be present in Cairo compiler repository root.
+        .map(|p| p.join("corelib"))
+    // Note, that for registry source, we do not get whole Cairo repository in cache.
+    // Thus the corelib will not be found - only the crate is downloaded.
```
[the remark this answers is not in the document]
Comment (maciektr) [reply utils/scarb-build-metadata/build.rs:118]: Oh yes, sorry!

File: utils/scarb-build-metadata/src/lib.rs [hunk 3]
```diff
@@ -25,7 +25,14 @@ pub const CAIRO_COMMIT_HASH: Option<CommitHash> = match (
...
     _ => panic!("Either SCARB_CAIRO_COMMIT_HASH or SCARB_CAIRO_SHORT_COMMIT_HASH is missing."),
 };
+
+// Commit hash corresponding to cairo compiler semver version.
 pub const CAIRO_COMMIT_REV: &str = env!("SCARB_CAIRO_COMMIT_REV");
+// Optional path to corelib from local cargo cache.
+//
+// If `None`, that the corelib should be attempted to be downloaded from Cairo
+// repository on GitHub.
+pub const SCARB_CORELIB_LOCAL_PATH: Option<&str> = option_env!("SCARB_CORELIB_LOCAL_PATH");
```
Comment (mkaput) [inline utils/scarb-build-metadata/src/lib.rs:35]:
```suggestion
/// Commit hash corresponding to cairo compiler semver version.
pub const CAIRO_COMMIT_REV: &str = env!("SCARB_CAIRO_COMMIT_REV");
/// Optional path to corelib from local cargo cache.
///
/// If `None`, that the corelib should be attempted to be downloaded from Cairo
/// repository on GitHub.
pub const SCARB_CORELIB_LOCAL_PATH: Option<&str> = option_env!("SCARB_CORELIB_LOCAL_PATH");
```

File: scarb/build.rs [hunk 4]
```diff
@@ -28,61 +29,77 @@ fn download_core(rev: &str) {
...
         if let Ok(cairo_archive) = env::var("CAIRO_ARCHIVE") {
+            // If `CAIRO_ARCHIVE` env variable is specified, prefer it.
             // Copy archive to `cairo_zip`, without keeping file attributes.
             eprintln!("Copying Cairo archive from `CAIRO_ARCHIVE={cairo_archive}`.");
             let mut src = File::open(&cairo_archive).unwrap();
             let mut dst = File::create(&cairo_zip).unwrap();
             io::copy(&mut src, &mut dst).unwrap();
+            extract_zipped_core(&cairo_zip, &core_path);
+        } else if let Some(corelib_path) = SCARB_CORELIB_LOCAL_PATH {
+            // Else, if corelib is present in Cargo cache, prefer it over remote.
```
Comment (mkaput) [inline scarb/build.rs:44]:
```suggestion
            // Otherwise; if corelib is present in Cargo cache, prefer it over remote.
```

**Review by mkaput — APPROVED** [review]

**Pull request merged**